### DISCUSSION TITLE
Disallow concurrent execution by IJob (but not by JobKey)

### DISCUSSION THREAD

**Discussion by Wojtekxor2** [quartznet/quartznet#1210] [General] [answered]
I wanted to make sure that only one instance is run by Quartz no matter what triggered it.

DisallowConcurrentExecution attribute seems to work only per JobKey, however when job is triggered from the scheduler and e.g. using RunJobOutOfSchedule has different job name, different trigger name and so on ...

I wrote something like this to prevent paraller execution: 

```
logger.LogInformation("Started");
var jobs = context.Scheduler.GetCurrentlyExecutingJobs().Result;
foreach (var job in jobs)
{
  var jobDetail = job.JobDetail as Quartz.Impl.JobDetailImpl;
  var currentJobDetail = context.JobDetail as Quartz.Impl.JobDetailImpl;

  if (jobDetail == null || currentJobDetail == null)
    return;

  if (jobDetail.Name.Equals(currentJobDetail.Name) && (!job.JobInstance.Equals(this)))
  {
    logger.LogInformation("There's another instance running");
    return;
  }
}

await Job();
```

Is there a better way to achieve the same?

**Comment by lahma** [comment, marked as answer]
Job's identity determines the uniqueness (instance) so you should share the job definition between the triggers, like you've noticed. I'm afraid there's no way around that. Is there are reason not to register the job definition only once and shared between competing triggers?

**Reply by Wojtekxor2** [reply]
As far as I know, JobKey is a combination of Group and Name. 

When I run OutOfSchedule I've got: Group: **OneTime** and Name: **MY_JOB_NAME**
When It is run from scheduler I've got: Group: **MY_GROUP_NAME** and Name:  **MY_JOB_NAME**

When that happens IJob might run concurrently. There is only one job definition: **MY_JOB_NAME** that's why I didn't understand the question: "Is there are reason not to register the job definition only once(...)"

What I wanted to achieve is only one IJob run at a time no matter who calls it from where.

**Reply by lahma** [reply]
What creates such group `OneTime`? Generally add the job once to scheduler then you can trigger it from many triggers referencing the same one and only job key. I don't know what this `RunJobOutOfSchedule`, but you can use `IScheduler.TriggerJob(the-one-and-only-job-key)` to schedule one-time invocation of jobs,

**Reply by Wojtekxor2** [reply]
Aaah, sorry. I didn't realize that method was part of my infrastructure.... No more questions ... :)
